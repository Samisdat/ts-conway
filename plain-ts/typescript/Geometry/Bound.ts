export class Bound {

    private min: number;

    private max: number;

    constructor(min: number, max: number) {

        if (max < min) {
            throw new Error('max can not be lower then min');
        }

        this.min = min;
        this.max = max;
    }

    public isWithin(test: number): boolean {
        return (test >= this.min && test <= this.max);
    }

    public isBelow(test: number): boolean {
        return (test < this.min);
    }

    public isAbove(test: number): boolean {
        return (test > this.max);
    }

    public confine(value: number): number {

        if (value < this.min) {
            return this.min;
        }

        if (value > this.max) {
            return this.max;
        }

        return value;

    }

    public getMax(): number{
        return this.max;
    }

    public getMin(): number{
        return this.min;
    }

    public expand(value: number): void{

        if(true === this.isWithin(value)){
            return;
        }

        if(true === this.isBelow(value)){
            this.min = value
        }

        if(true === this.isAbove(value)){
            this.max = value
        }

    }

}
