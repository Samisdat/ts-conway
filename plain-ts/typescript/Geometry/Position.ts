export class Position {

    public readonly  x: number;
    public readonly  y: number;

    constructor(x: number, y: number) {

        if (-0 === x) {
            x = 0;
        }

        if (-0 === y) {
            y = 0;
        }

        this.x = x;
        this.y = y;
    }

    public move(position: Position): Position {
        return new Position(
            this.x + position.x,
            this.y + position.y
        );
    }

    public inverse(): Position {
        return new Position(
            -1 * this.x,
            -1 * this.y
        );
    }

    public compare(position: Position): boolean {

        return (position.x === this.x && position.y === this.y);

    }

    public getNeighbours(): Position[] {

        const neighbours: Position[] = [];

        // neighbours on bottom
        neighbours.push(
            new Position(this.x - 1, this.y - 1)
        );
        neighbours.push(
            new Position(this.x, this.y - 1)
        );
        neighbours.push(
            new Position(this.x + 1, this.y - 1)
        );

        // neighbours on the left and on the right
        neighbours.push(
            new Position(this.x - 1, this.y)
        );
        neighbours.push(
            new Position(this.x + 1, this.y)
        );

        // neighbours on top
        neighbours.push(
            new Position(this.x - 1, this.y + 1)
        );
        neighbours.push(
            new Position(this.x, this.y + 1)
        );
        neighbours.push(
            new Position(this.x + 1, this.y + 1)
        );

        return neighbours;
    }

    public clone(): Position {
        return new Position(
            this.x,
            this.y
        );
    }

    public toString(): string {

        let str = '';

        str += (0 <= this.x) ? '+' : '';
        str += this.x;

        str += '/';

        str += (0 <= this.y) ? '+' : '';
        str += this.y;

        return str;

    }
}
