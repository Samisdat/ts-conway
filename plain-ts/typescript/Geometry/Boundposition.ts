import {Position} from '@Conway/Geometry/Position';
import {Bound} from '@Conway/Geometry/Bound';

export class Boundposition {

    private x: Bound;
    private y: Bound;

    constructor(topLeft: Position, bottomRight: Position) {

        if (bottomRight.x < topLeft.x) {
            throw new Error('bottomRight must be right from topLeft');
        }

        if (bottomRight.y < topLeft.y) {
            throw new Error('bottomRight must be below from topLeft');
        }

        this.x = new Bound(topLeft.x, bottomRight.x);
        this.y = new Bound(topLeft.y, bottomRight.y);

    }

    public isWithin(test: Position): boolean {
        return (this.x.isWithin(test.x) && this.y.isWithin(test.y));
    }

    public confine(value: Position): Position {

        return new Position(
            this.x.confine(value.x),
            this.y.confine(value.y)
        );

    }

    public bottomLeft(): Position{

        return new Position(this.x.getMin(), this.y.getMin());

    }

    public topRight(): Position{

        return new Position(this.x.getMax(), this.y.getMax());

    }

    public getWidth(): number{

        return this.x.getMax() - this.x.getMin() + 1;

    }

    public getHeight(): number{

        return this.y.getMax() - this.y.getMin() + 1;

    }

    public expand(position: Position): void{

        if(true === this.isWithin(position)){
            return;
        }

        this.x.expand(position.x);
        this.y.expand(position.y);

    }

}
