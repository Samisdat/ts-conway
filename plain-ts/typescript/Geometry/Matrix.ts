import {Position} from '@Conway/Geometry/Position';
import {Boundposition} from '@Conway/Geometry/Boundposition';

interface MatrixInterface {
    [index: string]: Position;
}


export class Matrix {

    private matrix: MatrixInterface = {};

    private boundposition: Boundposition;

    public has(position: Position): boolean {

        return (undefined !== this.matrix[position.toString()]);

    }

    public add(position: Position): void {

        if (true === this.has(position)) {
            return;
        }

        this.matrix[position.toString()] = position;


        if(undefined === this.boundposition){
            this.boundposition = new Boundposition(position, position);
        }
        else{
            this.boundposition.expand(position);
        }

    }

    public remove(position: Position): void {

        if (false === this.has(position)) {
            return;
        }

        delete this.matrix[position.toString()];

    }

    public getAll(): Position[] {

        const positions: Position[] = [];

        for (const index in this.matrix) {

            positions.push(this.matrix[index]);

        }

        return positions;

    }

    public getBound(): Boundposition {

        return this.boundposition;

    }

    public getWidth(): number {

        return this.getBound().getWidth();

    }

    public getHeight(): number {

        return this.getBound().getHeight();

    }

}
