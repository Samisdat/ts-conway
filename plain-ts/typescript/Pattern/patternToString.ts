import {Pattern} from '@Conway/Pattern/Pattern';
import {matrixToString} from '@Conway/Geometry/matrixToString';

export const patternToString = (pattern: Pattern): string => {

    let patternAsString = '!Name: ' + pattern.getName() + '\n';

    patternAsString = patternAsString + matrixToString(pattern.getMatrix(), false)

    return patternAsString;

}
