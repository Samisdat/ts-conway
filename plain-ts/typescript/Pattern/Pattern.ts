import {Position} from '@Conway/Geometry/Position';
import {readPatternFromPlainFile} from '@Conway/Pattern/readPatternFromPlainFile';
import {Matrix} from '@Conway/Geometry/Matrix';
import {patternToString} from '@Conway/Pattern/patternToString';

export class Pattern {

    private name: string;

    private matrix: Matrix;

    constructor(name: string, matrix: Matrix) {

        this.name = name;

        this.matrix = matrix;

    }

    public static fromString(patternString: string): Pattern {
        return readPatternFromPlainFile(patternString);
    }

    public getName(): string {
        return this.name;
    }

    public getMatrix(): Matrix {

        return this.matrix;
    }

    public getWidth(): number {
        return this.matrix.getWidth();
    }

    public getHeight(): number {
        return this.matrix.getHeight();
    }

    public mirrorVertical(): void {

        const mirror = new Matrix();

        for (const position of this.matrix.getAll()) {

            mirror.add(
                new Position(
                    position.x,
                    -1 * position.y
                )
            );

        }

        this.matrix = mirror;

    }

    public mirrorHorizontal(): void {

        const mirror = new Matrix();

        for (const position of this.matrix.getAll()) {

            mirror.add(
                new Position(
                    -1 * position.x,
                    position.y
                )
            );

        }

        this.matrix = mirror;

    }

    public rotate(): void {

        const matrix = this.toArray();

        const rotated: number[][] = [];

        for (let x = 0; x < matrix[0].length; x += 1) {

            const row: number[] = [];

            for (let y = matrix.length - 1; y >= 0 ; y -= 1) {

                row.push(matrix[y][x]);

            }

            rotated.push(row);

        }
        const rotatedMatrix = new Matrix();

        for (let y = 0; y < rotated.length; y += 1) {

            for (let x = 0; x < rotated[y].length; x += 1) {

                if (0 === rotated[y][x]) {
                    continue;
                }

                rotatedMatrix.add(new Position(x, y));

            }

        }

        this.matrix = rotatedMatrix;

    }

    public toArray(): number[][] {

        const moved = new Matrix();

        const moveBy = this.getMatrix().getBound().bottomLeft().inverse();

        for (const position of this.matrix.getAll()) {

            moved.add(
                position.move(moveBy)
            );

        }

        const matrixAsArray: number[][] = [];

        for (let y = 0; y < this.getHeight() + 1; y += 1) {

            const row: number[] = [];

            for (let x = 0; x < this.getWidth() + 1; x += 1) {

                const cell: number = (true === moved.has(new Position(x, y))) ? 1 : 0;

                row.push(cell);

            }

            matrixAsArray.push(row);
        }

        return matrixAsArray;

    }

    public toString(): string {

        return patternToString(this);

    }

}
