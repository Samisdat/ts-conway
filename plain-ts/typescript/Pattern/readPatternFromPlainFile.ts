import {Position} from '@Conway/Geometry/Position';
import {Pattern} from '@Conway/Pattern/Pattern';
import {Matrix} from '@Conway/Geometry/Matrix';

interface ExtractedPattern {
    name: string;
    pattern: string[];
}

const extractePattern = (patternPlainText: string): ExtractedPattern => {

    const extractedPattern: ExtractedPattern = {
        name: 'unnamed',
        pattern: []
    };

    const rows: string[] = patternPlainText.split('\n').reverse();

    const isComment = new RegExp('^!(.*?)');
    const isMeta = new RegExp('^!(.*?)\:(.*?)$');

    for (const row of rows) {

        if (true === isComment.test(row)) {

            if (true === isMeta.test(row)) {

                const meta = isMeta.exec(row) as RegExpMatchArray;

                if ('Name' === meta[1]) {
                    extractedPattern.name = meta[2].trim();
                }

            }

        } else {

            extractedPattern.pattern.push(row);

        }

    }

    return extractedPattern;

};

export const readPatternFromPlainFile = (patternPlainText: string): Pattern => {

    const extractedPattern = extractePattern(patternPlainText);

    const matrix = new Matrix();

    let y = 0;
    let x = 0;

    for (const row of extractedPattern.pattern) {

        const cols: string[] = row.trim().split('');

        x = 0;

        for (const col of cols) {

            if ('O' === col) {

                matrix.add(new Position(x, y));

            }

            x += 1;
        }

        y += 1;

    }

    const pattern = new Pattern(extractedPattern.name, matrix);

    return pattern;

};

