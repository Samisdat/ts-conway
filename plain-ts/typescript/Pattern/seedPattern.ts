
import {Pattern} from '@Conway/Pattern/Pattern';
import {Position} from '@Conway/Geometry/Position';
import {Matrix} from '@Conway/Geometry/Matrix';

export const seedPattern = (matrix: Matrix, pattern: Pattern, moveBy = new Position(0, 0)): void => {

    const positions = pattern.getMatrix().getAll();

    const offSet = new Position(
        -1 * Math.floor(pattern.getMatrix().getWidth() / 2),
        -1 * Math.floor(pattern.getMatrix().getHeight() / 2)
    ).move(moveBy);

    for (const position of positions) {

        matrix.add(position.move(moveBy));

    }

}
