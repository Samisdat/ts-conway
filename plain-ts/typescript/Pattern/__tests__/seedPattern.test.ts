import serializer from '../../../jest-serialize-conway';
import {Position} from '@Conway/Geometry/Position';
import {Pattern} from '@Conway/Pattern/Pattern';
import {blinker} from '@Conway/Pattern/Store/blinker.cells';
import {seedPattern} from '@Conway/Pattern/seedPattern';
import {Matrix} from '@Conway/Geometry/Matrix';

describe('seedPattern', () => {

    beforeEach(function () {
        expect.addSnapshotSerializer(serializer);
    });

    test('seed a pattern', () => {

        const patternString = blinker

        const pattern = Pattern.fromString(blinker);

        const matrix = new Matrix();

        seedPattern(matrix, pattern);

        expect(pattern.getMatrix()).toMatchSnapshot();

    });

    test('seed a pattern with offset', () => {

        const patternString = blinker

        const pattern = Pattern.fromString(blinker);

        const matrix = new Matrix();

        seedPattern(matrix, pattern, new Position(1,1));

        expect(pattern.getMatrix()).toMatchSnapshot();

    });

});
