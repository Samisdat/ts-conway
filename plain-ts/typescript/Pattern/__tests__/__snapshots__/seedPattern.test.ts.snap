// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`seedPattern seed a pattern 1`] = `
XXX
OOO
OOO
`;

exports[`seedPattern seed a pattern with offset 1`] = `
XXX
OOO
OOO
`;
